### DISCUSSION TITLE
Cocos2dx 4.0 to Axmol migration questions

### DISCUSSION THREAD

**Discussion by IcemarkUK** [axmolengine/axmol#1660] [Q&A] [answered]
So - I have finally started migrating my project and I have the it up and running - at least to the main menu.

A few questions on what is not working a the moment...

FYI: I'm just testing the OSX build at the moment.

#### axmol build

I couldn't get this to work because I couldn't pass in variables to cmake. I use the following currently
`cmake -GXcode -DTME:string=LOM -B Builds/mac-build-lom`

Because I use the same project to generate different variants.

I tried
`axmol build -p osx -a x64 -xc '-DTME:string=LOM'`

but xc always ended up as empty

```
Name                           Value
----                           -----
sdk                            
d                              /Users/chris/Projects/GitHub/The-Lords-Of-Midnight/main/midnight
xb                             {--target, midnight}
cc                             
prefix                         /Users/chris/Developer/axmol/tools/external
a                              x64
t                              
dll                            False
p                              osx
xt                             cmake
xc                             {}

```
#### TransitionCrossFade
~~This used to cross fade using black. It now appears to be a blue-ish colour~~
The Cross Fade isn't picking up the target scene until the end. Thus it has a blueish colour of an empty scene.

#### Shaders
I only have 2 shaders. And they are crashing in ShaderModuleMTL::ShaderModuleMTL when I do
    program = cocos2d::backend::DriverBase::getInstance()->newProgram(vertSource, fragSource);

I presume the shader needs converting to something else?    I've including them here for reference.

[characterTimeShader.fsh.txt](https://github.com/axmolengine/axmol/files/14216325/characterTimeShader.fsh.txt)
[standard.vsh.txt](https://github.com/axmolengine/axmol/files/14216326/standard.vsh.txt)
[terrainTimeShader.fsh.txt](https://github.com/axmolengine/axmol/files/14216327/terrainTimeShader.fsh.txt)

#### Libraries
How do I turn off all the thirdpary libraries, extensions, and features that I don't need. I never did this Cocos2dx but I figured it would be useful as there seems to be an awful lot of things that I don't need.

Thanks!

**Comment by danialias** [comment]
Hi! You'll find the answer to the last question in our Wiki: https://github.com/axmolengine/axmol/wiki/FAQ

**Comment by rh101** [comment]
> I only have 2 shaders. And they are crashing in ShaderModuleMTL::ShaderModuleMTL when I do program = cocos2d::backend::DriverBase::getInstance()->newProgram(vertSource, fragSource);
> 
> I presume the shader needs converting to something else? I've including them here for reference.
> 
> [characterTimeShader.fsh.txt](https://github.com/axmolengine/axmol/files/14216325/characterTimeShader.fsh.txt) [standard.vsh.txt](https://github.com/axmolengine/axmol/files/14216326/standard.vsh.txt) [terrainTimeShader.fsh.txt](https://github.com/axmolengine/axmol/files/14216327/terrainTimeShader.fsh.txt)

I'll try to convert these for you and post them in here.  They can possibly serve as a reference for anyone else trying to convert shaders too.

> `axmol build -p osx -a x64 -xc '-DTME:string=LOM'`
> 
> but xc always ended up as empty

I tried this on Windows, and while xc isn't empty in the output as it is for you, I don't see the reference in the generated vcxproj file, so there may be something wrong with the `axmol build` command.

**Comment by rh101** [comment]
@IcemarkUK Try the converted shaders in this zip:

[converted_shaders.zip](https://github.com/axmolengine/axmol/files/14219822/converted_shaders.zip)

Note - make sure the shaders are placed in `[PROJECT_DIR]/Source/shaders` so that they're processed correctly.

Example of usage based on one of your shaders:
```
auto* program = ProgramManager::getInstance()->loadProgram(vertexShaderPath, fragmentShaderPath);
// program should NOT be released, even though it is not an auto-release object,
// since it is cached (otherwise we would accidentally be releasing the cached entry)
auto* programState = new ProgramState(program);

auto colourLoc = programState->getUniformLocation("p_colour");
auto alphaLoc = programState->getUniformLocation("p_alpha");
auto fadeLoc = programState->getUniformLocation("p_fade");

node->setProgramState(programState, [true | false]); // ownPS = true or false .  If true, then the node will not call retain on the programState, and if false, it will.  Default is false.

// Updating values in the shader
programState->setUniform(colourLoc, &_colour, sizeof(_colour));  // Vec4 _colour
programState->setUniform(alphaLoc, &_alpha, sizeof(_alpha)); // float _alpha
programState->setUniform(fadeLoc, &_fade, sizeof(_fade)); // float _fade
```

**Reply by IcemarkUK** [reply]
Many thanks for doing that...

When you say the shaders should be in [PROJECT_DIR]/Source/shaders - previously I had them in [PROJECT_DIR]/Resources/shaders

I was loading them with
    auto fragSourceRaw = cocos2d::FileUtils::getInstance()->getStringFromFile("shaders/characterTimeShader.fsh");

I presume if they are in the source folder, then they need an entry in the CMakeLists to get them in to the resources folder?

**Reply by rh101** [reply]
> Many thanks for doing that...

No problem.

> When you say the shaders should be in [PROJECT_DIR]/Source/shaders - previously I had them in [PROJECT_DIR]/Resources/shaders
> 
> I was loading them with auto fragSourceRaw = cocos2d::FileUtils::getInstance()->getStringFromFile("shaders/characterTimeShader.fsh");

The shaders in `Source/shaders` should not longer be placed in the `Contents` directory (previously named `Resources`) as they are, since they need to be processed by a utility at build time in order to convert them into the correct shader format for the target platform (OpenGL/Metal etc.).  Once they're processed, the new files are saved in `[build_dir]/runtime/axslc/custom`.  The `axslc` dir is then copied over to your binary path, and that path is automatically added to the search list in Axmol.

So, in your own code, the path you reference is `custom/shadername`.  The shaders will have extensions `_fs` and `_vs`.  For your shaders,  the paths you need to use will be:

`custom/characterTimeShader_fs`
`custom/standard_vs`
`custom/terrainTimeShader_fs`

Note that the shaders that are part of the Axmol engine are located in `axslc`, so you reference them without the `custom` sub-dir.  Refer to the file `/core/renderer/Shaders.cpp` to see what I mean.

For example:
`auto* program = ProgramManager::getInstance()->loadProgram("custom/standard_vs", "custom/characterTimeShader_fs");`

> I presume if they are in the source folder, then they need an entry in the CMakeLists to get them in to the resources folder?

No need, it's all automated.  The only reason you would want entries in `CMakeLists.txt` is if you're storing your shaders in a directory other than the one the build scripts expect, being `Source/shaders`.

**Reply by IcemarkUK** [reply]
So just as an aside - my source folder was /src/ and my resources /resources/ I presumed that just changing the references in the originally created CMakeFiles.txt would suffice - I have CMakeFiles.txt throughout my src directory so I removed the default catchall recursive one. If I was to leave src/ what's the best way to make the shaders get picked up correctly?

**Reply by IcemarkUK** [reply]
Thanks Shaders sorted.

**Reply by rh101** [reply]
> So just as an aside - my source folder was /src/ and my resources /resources/ I presumed that just changing the references in the originally created CMakeFiles.txt would suffice - I have CMakeFiles.txt throughout my src directory so I removed the default catchall recursive one. If I was to leave src/ what's the best way to make the shaders get picked up correctly?

You shouldn't be able to have the folder structure as you want, and just set everything in `CMakeListst.txt`.  Best thing to do would be to search `*.txt` and `*.cmake` within the engine source to see where the paths are being used.  I don't have time at the moment to look into it, but perhaps later today I can (in approx 10 hours).  

If for some reason the paths are hard-coded into the engine, then that should be something to change.  The engine shouldn't be assuming the file structure of the game project.

EDIT: Unfortunately, the paths are hard-coded in file `AXBuildHelpers.cmake`, function `ax_setup_app_config`.  I feel that it should be changed to allow the game project to set the paths, as the engine scripts shouldn't be assuming the location of both source and contents.

**Reply by IcemarkUK** [reply]
Thanks for checking - I've renamed my folders for now. I was just trying to keep a lower commit count down for the migration - but it's not a big deal.

**Comment by halx99** [comment, marked as answer]
please use `axmol build -p osx -a x64 -xc '-DTME=LOM'` because powershell cmdlet arg interpreter will divide arg value by `:`

**Reply by IcemarkUK** [reply]
Can you change the build directory from being in this case 'Build_x64' eg. 'Builds/Build_x64_Lom'  I am building 3 projects with this command... I could rename the folder after build - but I wondered if there is a command line for it.

**Reply by halx99** [reply]
pass -B like:

`axmol build -p osx -a x64 -xc '-DTME=LOM,-Bbuild_lom'`

**Reply by IcemarkUK** [reply]
Thanks
